engine/src/main/java/org/terasology/i18n/TranslationSystemImpl.java

```diff
@@ -90,31 +77,72 @@ public TranslationProject getProject(Uri name) {
     }
 
     @Override
-    public String translate(String id) {
-        return translate(id, config.getLocale());
+    public String translate(String id, Object... arguments) {
+        return translate(id, config.getLocale(), arguments);
     }
 
     @Override
-    public String translate(String text, Locale otherLocale) {
-        TemplateEngine templateEngine = new TemplateEngineImpl(id -> {
-            ResourceUrn uri = new ResourceUrn(id);
-            SimpleUri projectUri = new SimpleUri(uri.getModuleName(), uri.getResourceName());
-            TranslationProject project = getProject(projectUri);
-            if (project != null) {
-                Optional<String> opt = project.translate(uri.getFragmentName(), otherLocale);
-                if (opt.isPresent()) {
-                    return opt.get();
-                } else {
-                    logger.warn("No translation for '{}'", id);
-                    return "?" + uri.getFragmentName() + "?";
-                }
+    public String translate(String text, Locale otherLocale, Object... arguments) {
+        Object[] expandedArguments = expandFormatArguments(otherLocale, arguments);
+        TemplateEngine templateEngine = new TemplateEngineImpl(id ->
+                translateFragment(otherLocale, id, expandedArguments));
+
+        return templateEngine.transform(text);
+    }
+
+    /**
+     * Expand string arguments as if they are translatable templates.
+     *
+     * Leaves non-string arguments as they are.
+     * <p>
+     * One level of expansion only; String arguments may not require arguments of their own.
+     * <p>
+     * Example:
+     * <pre>{@code
+     *     Object[] arguments = {17, "${engine:menu#credits}", new Date(16e11)};
+     *     Object[] expandedArguments = expandFormatArguments(locale, arguments);
+     *
+     *     println(expandedArguments)  // {17, "Credits", Date(16e11)}
+     * }</pre>
+     *
+     * @param otherLocale FIXME: why <em>is</em> this "other" locale and not just "locale"?
```

Comment:
Probably the class had a member called `locale` at some point and Checkstyle would warn if you'd overload the name in a non-setter context? 🤷‍♂️